packages/ra-core/src/dataProvider/useQuery.ts

```diff
@@ -113,19 +123,21 @@ const useQuery = (
                     : [payload, otherOptions]
             )
             .then(({ data, total }) => {
-                setState({
+                setState(prev => ({
+                    ...prev,
```

Comment:
I remember that we used to do that but we had problems in corner cases. I suggest you pass `refetch` explicitly  instead.